# Setup and Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set
import warnings
import re
from pandas.io import gbq
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from scipy import sparse

# Get the Data

To get the data, we'll write a function that fetches the data from Google BigQuery. We'll create a Python Module that will hold the code below to be used in other notebooks.

In [2]:
def fetch_chicago_data(query_string, project_id, excel_file, verbose = True):
    """
    This function fetches the data from Google BigQuery. First a query is made based on the specified query string and
    project id. Next, an excel file with the Chicago districts are read in and joined with the original data. Finally, 
    a duplicate column is dropped.
    
    query_string: String with the specific query to be passed to Google BigQuery
    project_id: Corresponding project id in Google BigQuery
    excel_file: Path to the excel file containing chicago districts
    
    returns: pandas dataframe with the queried and joined data
    """
    
    if verbose: print("Fetching Chicago Data Started...\n")
        
    #Read the data in from Google BigQuery
    chicago_data = gbq.read_gbq(query_string, project_id=project_id)
    if verbose: print("Successfully queried Google BigQuery.")
        
    #Read in an Excel file with a one to one mapping between Chicago community areas and districts
    chicago_districts = pd.read_excel(excel_file)
    if verbose: print("Sucessfully read in excel file.")
        
    #Data type can't be joined on an int
    chicago_districts.community_area = chicago_districts["community_area"].astype("string")
    chicago_data.community_area = chicago_data["community_area"].astype("string")
        
    #Outer join the two data sets
    chicago = chicago_data.merge(chicago_districts, how="outer", left_on="community_area", right_on="community_area")
    if verbose: print("Sucessfully joined Chicago districts to main data.")
    
    #Drop the community area variable since we have a community name variable
    chicago.drop("community_area", axis = 1, inplace = True)
    if verbose: print("Successfully dropped duplicate column")
        
    if verbose: print("\nSucccessfully fetched Chicago Data")
        
    return chicago

In [3]:
%%time
#Specify input values for fetching the data
query = """
            SELECT unique_key, date, primary_type, location_description, 
                    arrest, domestic, community_area, year
            FROM `gdac-327115.Chicago.chicago2`
            WHERE year >= 2011
        """
project_id = "gdac-327115"
excel_file = "ChicagoCommunityAreas.xlsx"

#Fetch the data
chicago = fetch_chicago_data(query, project_id, excel_file, verbose=True)

Fetching Chicago Data Started...

Successfully queried Google BigQuery.
Sucessfully read in excel file.
Sucessfully joined Chicago districts to main data.
Successfully dropped duplicate column

Succcessfully fetched Chicago Data
Wall time: 3min 39s


In [4]:
#Check data types
chicago.dtypes

unique_key                            Int64
date                    datetime64[ns, UTC]
primary_type                         object
location_description                 object
arrest                              boolean
domestic                            boolean
year                                  Int64
district_name                        object
community_name                       object
dtype: object

In [9]:
#Change the object data types to categorical
chicago.loc[:, chicago.dtypes == "object"] = chicago.select_dtypes(["object"]).apply(lambda x: x.astype("category"))

### Explore first few rows

In [10]:
chicago.head()

unique_key                      date            primary_type  \
0    11228094 2017-11-20 15:00:00+00:00      DECEPTIVE PRACTICE   
1    12135292 2020-08-13 12:00:00+00:00      DECEPTIVE PRACTICE   
2     8905294 2012-11-27 17:30:00+00:00  PUBLIC PEACE VIOLATION   
3     8805120 2012-09-17 00:30:00+00:00      DECEPTIVE PRACTICE   
4     9030552 2013-02-21 00:00:00+00:00     CRIM SEXUAL ASSAULT   

  location_description  arrest  domestic  year   latitude  longitude  \
0            POOL ROOM   False     False  2017        NaN        NaN   
1                 BANK   False     False  2020  41.993402 -87.712280   
2           RESTAURANT   False     False  2012  41.978265 -87.708742   
3              TAXICAB   False     False  2012  41.985277 -87.713827   
4            RESIDENCE   False      True  2013  41.994323 -87.720213   

  district_name community_name  
0     Far North     NORTH PARK  
1     Far North     NORTH PARK  
2     Far North     NORTH PARK  
3     Far North     NORTH PARK  
4     Far North     NORTH PARK

In [11]:
#Check the number of rows and columns
chicago.shape

(2971810, 11)

# Split the Data into a Training Set and Test Set

In [10]:
chicago_train = chicago.loc[chicago["year"] != 2021]
chicago_test = chicago.loc[chicago["year"] == 2021]

In [20]:
#Check if the test set contains data from the full year
chicago_test.loc[:, "date"].dt.month. \
                                value_counts(). \
                                reset_index(). \
                                rename(columns={"index":"Month", "date":"Count"}). \
                                sort_values(by = "Month")

Month  Count
5      1  16008
8      2  12852
6      3  15709
7      4  15279
4      5  17493
2      6  18468
0      7  18898
3      8  18110
1      9  18632
9     10   5507

# Data Exploration and Cleaning

In [22]:
#Check the total number of missing values for each variable
chicago_train.isnull().sum()

unique_key                  0
date                        0
primary_type                0
location_description     7844
arrest                      0
domestic                    0
year                        0
latitude                26843
longitude               26843
district_name             218
community_name            218
dtype: int64

In [21]:
#Check to make sure there are no duplicate incidents based on the "unique key" variable
assert len(chicago_train["unique_key"].unique()) == chicago_train.shape[0]

### Cleaning Primary Type Variable

In [23]:
#Quick reference on the number of missing values
chicago_train["primary_type"].isnull().sum()

0

In [10]:
#Examine the different crime types listed
chicago_train["primary_type"].value_counts()

THEFT                                636007
BATTERY                              512318
CRIMINAL DAMAGE                      299943
NARCOTICS                            222406
ASSAULT                              189549
OTHER OFFENSE                        171368
DECEPTIVE PRACTICE                   167709
BURGLARY                             152532
MOTOR VEHICLE THEFT                  120013
ROBBERY                              108092
CRIMINAL TRESPASS                     69971
WEAPONS VIOLATION                     45871
OFFENSE INVOLVING CHILDREN            23359
PUBLIC PEACE VIOLATION                21829
PROSTITUTION                          12438
CRIM SEXUAL ASSAULT                   12334
INTERFERENCE WITH PUBLIC OFFICER      11793
SEX OFFENSE                           10771
HOMICIDE                               5669
ARSON                                  4482
GAMBLING                               3507
LIQUOR LAW VIOLATION                   3407
CRIMINAL SEXUAL ASSAULT         

It appears "CRIM SEXUAL ASSAULT" should be merged into "CRIMINAL SEXUAL ASSAULT". We'll also merge the three various "Non-criminal" categories into one category. We'll do the same thing for the "OTHER NARCOTIC VIOLATION" category. Finally, we'll move "RITUALISM" into the "OTHER" since it has only one occurence. Let's collect all of these changes into one function. 

In [11]:
def crime_type_cleaner(df):
    """
    This function cleans the primary_type column in place. 
    
    df: DataFrame with primary_type column
    
    returns: None
    """
    df.loc[:, "primary_type"].replace(["CRIM SEXUAL ASSAULT"], "CRIMINAL SEXUAL ASSAULT", inplace = True)

    df.loc[:, "primary_type"].replace(["OTHER NARCOTIC VIOLATION"], "NARCOTICS", inplace = True)

    df.loc[:, "primary_type"].replace(["NON - CRIMINAL", "NON-CRIMINAL (SUBJECT SPECIFIED)"], "NON-CRIMINAL", inplace = True)

    df.loc[:, "primary_type"].replace(["RITUALISM"], "OTHER OFFENSE", inplace = True)
   

In [29]:
#Clean the primary type variable
crime_type_cleaner(chicago_train)
#Check the cleaned variable
chicago_train["primary_type"].value_counts()

THEFT                                636007
BATTERY                              512318
CRIMINAL DAMAGE                      299943
NARCOTICS                            222467
ASSAULT                              189549
OTHER OFFENSE                        171369
DECEPTIVE PRACTICE                   167709
BURGLARY                             152532
MOTOR VEHICLE THEFT                  120013
ROBBERY                              108092
CRIMINAL TRESPASS                     69971
WEAPONS VIOLATION                     45871
OFFENSE INVOLVING CHILDREN            23359
PUBLIC PEACE VIOLATION                21829
CRIMINAL SEXUAL ASSAULT               14850
PROSTITUTION                          12438
INTERFERENCE WITH PUBLIC OFFICER      11793
SEX OFFENSE                           10771
HOMICIDE                               5669
ARSON                                  4482
GAMBLING                               3507
LIQUOR LAW VIOLATION                   3407
KIDNAPPING                      

### Cleaning Location Description Variable

In [11]:
#Check the number of missing values
chicago_train["location_description"].isnull().sum()

7844

In [12]:
#Look at which rows are missing
chicago_train.loc[chicago_train["location_description"].isnull()]

unique_key                      date        primary_type  \
2783       11405363 2018-03-16 08:10:00+00:00  DECEPTIVE PRACTICE   
2791       12046546 2020-05-08 09:30:00+00:00  DECEPTIVE PRACTICE   
6360       11376002 2017-07-19 09:50:00+00:00  DECEPTIVE PRACTICE   
6366       11515304 2018-11-23 00:35:00+00:00  DECEPTIVE PRACTICE   
6371       12175047 2020-09-22 11:55:00+00:00  DECEPTIVE PRACTICE   
...             ...                       ...                 ...   
2970637    11503719 2017-10-01 08:30:00+00:00  DECEPTIVE PRACTICE   
2970640    12491726 2017-11-30 12:00:00+00:00  DECEPTIVE PRACTICE   
2970654    12232285 2015-10-01 16:15:00+00:00  DECEPTIVE PRACTICE   
2970655    12195264 2020-10-09 10:25:00+00:00  DECEPTIVE PRACTICE   
2970657    11830766 2019-08-20 12:35:00+00:00  DECEPTIVE PRACTICE   

        location_description  arrest  domestic  year  district_name  \
2783                     NaN   False     False  2018  Far Southeast   
2791                     NaN   False     False  2020  Far Southeast   
6360                     NaN   False     False  2017  Far Southeast   
6366                     NaN   False     False  2018  Far Southeast   
6371                     NaN   False     False  2020  Far Southeast   
...                      ...     ...       ...   ...            ...   
2970637                  NaN   False     False  2017      Far North   
2970640                  NaN   False     False  2017      Far North   
2970654                  NaN   False     False  2015      Far North   
2970655                  NaN   False     False  2020      Far North   
2970657                  NaN   False     False  2019      Far North   

        community_name  
2783         RIVERDALE  
2791         RIVERDALE  
6360         RIVERDALE  
6366         RIVERDALE  
6371         RIVERDALE  
...                ...  
2970637      EDGEWATER  
2970640      EDGEWATER  
2970654      EDGEWATER  
2970655      EDGEWATER  
2970657      EDGEWATER  

[7844 rows x 9 columns]

Looks like there might be some relationship between the missing locations and the type of crime committed. Let's further explore this. 

In [13]:
#Look at the counts of the primary type variable with the data filtered to those values of location that are missing
chicago_train.loc[chicago_train["location_description"].isnull()]["primary_type"].value_counts()

DECEPTIVE PRACTICE                   7124
THEFT                                 704
BURGLARY                                5
ROBBERY                                 4
BATTERY                                 2
CRIMINAL DAMAGE                         2
OTHER OFFENSE                           1
ARSON                                   1
CRIMINAL SEXUAL ASSAULT                 1
SEX OFFENSE                             0
NON-CRIMINAL                            0
RITUALISM                               0
PUBLIC PEACE VIOLATION                  0
PUBLIC INDECENCY                        0
PROSTITUTION                            0
STALKING                                0
OTHER NARCOTIC VIOLATION                0
OFFENSE INVOLVING CHILDREN              0
OBSCENITY                               0
NON-CRIMINAL (SUBJECT SPECIFIED)        0
MOTOR VEHICLE THEFT                     0
NON - CRIMINAL                          0
NARCOTICS                               0
ASSAULT                           

It appears that nearly all missing values for location description have a primary type that is labeled as "DECEPTIVE PRACTICE". Let's see what the non-missing values for location description have for their corresponding "DECEPTIVE PRACTICE" values.

In [24]:
#Create a boolean mask that filters the data to non-missing locations with a primary type of "DECEPTIVE PRACTICE"
mask = (chicago_train["location_description"].notnull()) & (chicago_train["primary_type"] == "DECEPTIVE PRACTICE")
chicago_train.loc[mask]["location_description"].value_counts()

RESIDENCE                54361
APARTMENT                17232
OTHER                    16680
STREET                    8369
BANK                      6910
                         ...  
PORCH                        0
POOLROOM                     0
EXPRESSWAY EMBANKMENT        0
FARM                         0
YMCA                         0
Name: location_description, Length: 197, dtype: int64

Thus, it appears that the "RESIDENCE" location occurrs the most for non-missing "DECEPTIVE PRACTICE" crimes. Therefore, we can impute the missing values for location that have the corresponding crime as "DECEPTIVE PRACTICE" with "RESIDENCE".  

Let's do the same process for the "THEFT" crime

In [23]:
#Create a boolean mask that filters the data to non-missing locations with a primary type of "THEFT"
mask = (chicago_train["location_description"].notnull()) & (chicago_train["primary_type"] == "THEFT")
chicago_train.loc[mask]["location_description"].value_counts()

STREET                152284
RESIDENCE              58720
APARTMENT              41793
SMALL RETAIL STORE     35004
DEPARTMENT STORE       32695
                       ...  
ELEVATOR                   0
DRIVEWAY                   0
SCHOOL YARD                0
CTA PROPERTY               0
YMCA                       0
Name: location_description, Length: 197, dtype: int64

Now, it appears that the "STREET" location is the most occurring for non-missing "THEFT" crimes. Let's impute the missing values for location that have the corresponding crime as "THEFT" with "STREET". 

Finally, we'll collect the remaining crimes into one group due to their low counts and check what crime is most occurring for the group.

In [37]:
crime_type = ["BURGLARY", "ROBBERY", "BATTERY", "CRIMINAL DAMAGE", "ARSON", "CRIMINAL SEXUAL ASSAULT", "OTHER OFFENSE"]
mask = (chicago_train["primary_type"].isin(crime_type)) & (chicago_train["location_description"].notnull())
chicago_train.loc[mask, ["location_description"]].value_counts()

location_description
RESIDENCE               277616
APARTMENT               248008
STREET                  235210
SIDEWALK                125040
RESIDENCE-GARAGE         35372
                         ...  
PARKING LOT                  0
CLUB                         0
OFFICE                       0
NURSING HOME                 0
YMCA                         0
Length: 197, dtype: int64

Again, it looks like "RESIDENCE" is the most occuring location for the group of crimes when location is not missing. Thus, we can impute the missing location values for the group of crimes as "RESIDENCE". 

We'll collect all of these missing value imputations into one function

In [14]:
def location_imputer(df):
    """
    This function imputes missing values in the location_description column by setting the missing value to the value that 
    occurs most in the primary_type column in place
    
    df: DataFrame that contains the location_description column
    
    returns: None
    """
    mask = (df.loc[:, "primary_type"] == "DECEPTIVE PRACTICE") & (df.loc[:, "location_description"].isnull())
    df.loc[mask, ["location_description"]] = "RESIDENCE"
    
    mask = (df.loc[:, "primary_type"] == "THEFT") & (df.loc[:, "location_description"].isnull())
    df.loc[mask, ["location_description"]] = "STREET"
    
    crime_type = ["BURGLARY", "ROBBERY", "BATTERY", "CRIMINAL DAMAGE", "ARSON", "CRIMINAL SEXUAL ASSAULT", "OTHER OFFENSE"]
    mask = (df.loc[:, "primary_type"].isin(crime_type)) & (df.loc[:, "location_description"].isnull())
    df.loc[mask, ["location_description"]] = "RESIDENCE"
    

In [16]:
#Impute the missing values
location_imputer(chicago_train)
#Check that there are no missing values
chicago_train["location_description"].isnull().sum()

0

Now we can examine the different locations

In [21]:
#Take a look at the different categories in the location description variable
chicago_train["location_description"].value_counts()

STREET          638800
RESIDENCE       473271
APARTMENT       366599
SIDEWALK        277648
OTHER           101780
                 ...  
LAUNDRY ROOM         1
LAKE                 1
LAGOON               1
YMCA                 1
CHA ELEVATOR         0
Name: location_description, Length: 197, dtype: int64

Clearly, there are too many categories, 197, to be used in a model with most variables having very few occurrences. Let's see if we can't combine some of the similar values into one category. For instance, there are numerous different ways to break down Chicago Transit Authority (CTA) which can all be combined into one CTA category

In [51]:
chicago_train.loc[chicago_train["location_description"].str.contains("CTA")]["location_description"].value_counts()[:20]

CTA TRAIN                                    15906
CTA PLATFORM                                 12793
CTA BUS                                      11066
CTA BUS STOP                                  6331
CTA STATION                                   5767
CTA GARAGE / OTHER PROPERTY                   4863
CTA TRACKS - RIGHT OF WAY                      168
CTA PARKING LOT / GARAGE / OTHER PROPERTY      137
CTA "L" PLATFORM                                 6
CTA PROPERTY                                     3
CTA SUBWAY STATION                               1
CTA "L" TRAIN                                    1
POLICE FACILITY/VEH PARKING LOT                  0
POLICE FACILITY / VEHICLE PARKING LOT            0
PAWN SHOP                                        0
PARKING LOT/GARAGE(NON.RESID.)                   0
PARKING LOT / GARAGE (NON RESIDENTIAL)           0
ABANDONED BUILDING                               0
HOTEL/MOTEL                                      0
POOL ROOM                      

We'll do this process of combining categories by putting them all into one function

In [54]:
def location_cleaner(df):
    """
    This function cleans the location_description by merging categories together that are logically related in place.
    
    df: DataFrame that contains the location_description column
    
    returns: None
    """

    #Combine all values related to CTA
    CTA = ["CTA TRAIN", "CTA PLATFORM", "CTA BUS", "CTA BUS STOP", "CTA STATION", "CTA GARAGE / OTHER PROPERTY", 
           "CTA TRACKS - RIGHT OF WAY", "CTA PARKING LOT / GARAGE / OTHER PROPERTY", """CTA "L" PLATFORM""",
          "CTA PROPERTY", """CTA "L" TRAIN""", "CTA SUBWAY STATION"]
    df["location_description"].replace(CTA, "CTA", inplace = True)

    #Combine GANGWAY into ALLEY
    df["location_description"].replace(["GANGWAY"], "ALLEY", inplace = True)

    #Combine all values related to Sports/Clubs
    sports = ["ATHLETIC CLUB", "SPORTS ARENA/STADIUM", "SPORTS ARENA / STADIUM", "CLUB"]
    df["location_description"].replace(sports, "SPORTS", inplace = True)

    #Combine all values related to VACANT
    vacant = ["VACANT LOT/LAND", "VACANT LOT / LAND", "VACANT LOT"]
    df["location_description"].replace(vacant, "VACANT", inplace = True)

    #Combine the two Gas Station values
    df["location_description"].replace(["GAS STATION DRIVE/PROP."], "GAS STATION", inplace = True)

    #Combine all values related to Water/River/Lake
    water = ["LAKEFRONT/WATERFRONT/RIVERBANK", "POOL ROOM", "BOAT/WATERCRAFT", "LAKEFRONT / WATERFRONT / RIVERBANK", 
             "BOAT / WATERCRAFT", "RIVER BANK", "POOLROOM", "LAKE", "LAGOON"]
    df["location_description"].replace(water, "WATER", inplace = True)

    #Combine all values related to Banks/Credit Unions
    bank = ["ATM (AUTOMATIC TELLER MACHINE)", "CURRENCY EXCHANGE", "CREDIT UNION", "SAVINGS AND LOAN"]
    df["location_description"].replace(bank, "BANK", inplace = True)

    #Combine all values related to Roads/Streets/Highway
    street = ["HIGHWAY/EXPRESSWAY", "BRIDGE", "HIGHWAY / EXPRESSWAY", "EXPRESSWAY EMBANKMENT"]
    df["location_description"].replace(street, "STREET", inplace = True)

    #Combine all values related to Residence/Houses or items inside a House
    residence = ["RESIDENCE PORCH/HALLWAY", "RESIDENCE-GARAGE", "RESIDENCE - YARD (FRONT / BACK)", "RESIDENCE - PORCH / HALLWAY", 
           "RESIDENCE - GARAGE", "RESIDENTIAL YARD (FRONT/BACK)", "HOUSE", "ROOMING HOUSE", "PORCH", "YARD", "HALLWAY", "GARAGE",
          "VESTIBULE", "STAIRWELL", "BASEMENT", "LAUNDRY ROOM", "DRIVEWAY", "DRIVEWAY - RESIDENTIAL"]
    df["location_description"].replace(residence, "RESIDENCE", inplace = True)

    #Combine all values related to College
    college = ["COLLEGE/UNIVERSITY GROUNDS", "COLLEGE/UNIVERSITY RESIDENCE HALL", "COLLEGE / UNIVERSITY - GROUNDS", 
           "COLLEGE / UNIVERSITY - RESIDENCE HALL"]
    df["location_description"].replace(college, "COLLEGE", inplace = True)

    #Combine all values related to CTA into one CTA category
    veh = ["VEHICLE NON-COMMERCIAL", "VEHICLE-COMMERCIAL", "VEHICLE - OTHER RIDE SHARE SERVICE (E.G., UBER, LYFT)", 
           "VEHICLE - OTHER RIDE SERVICE", "VEHICLE - DELIVERY TRUCK", "VEHICLE - OTHER RIDE SHARE SERVICE (LYFT, UBER, ETC.)", 
           "VEHICLE - COMMERCIAL", "VEHICLE-COMMERCIAL - TROLLEY BUS", "VEHICLE-COMMERCIAL - ENTERTAINMENT/PARTY BUS", 
           "VEHICLE - COMMERCIAL: TROLLEY BUS", "VEHICLE - COMMERCIAL: ENTERTAINMENT / PARTY BUS", "TAXICAB", "",
          "AUTO / BOAT / RV DEALERSHIP", "AUTO", "DELIVERY TRUCK", "GARAGE/AUTO REPAIR", "TAXI CAB", "TRUCK", "TRAILER"]
    df["location_description"].replace(veh, "VEHICLE", inplace = True)

    #Combine all values related to Police/Fire Station/Protection
    police = ["POLICE FACILITY/VEH PARKING LOT", "POLICE FACILITY / VEHICLE PARKING LOT", "JAIL / LOCK-UP FACILITY",
             "FIRE STATION", "FOREST PRESERVE"]
    df["location_description"].replace(police, "POLICE", inplace = True)

    #Combine all values related to Parking
    parking = ["PARKING LOT/GARAGE(NON.RESID.)", "CHA PARKING LOT/GROUNDS", "PARKING LOT / GARAGE (NON RESIDENTIAL)", 
            "CHA PARKING LOT / GROUNDS", "PARKING LOT", "CHA PARKING LOT"]
    df["location_description"].replace(parking, "PARKING", inplace = True)

    #Combine all values related to Airports
    air = ["AIRPORT TERMINAL UPPER LEVEL - SECURE AREA", "AIRPORT TERMINAL LOWER LEVEL - NON-SECURE AREA", 
           "AIRPORT BUILDING NON-TERMINAL - NON-SECURE AREA", "AIRPORT VENDING ESTABLISHMENT", "AIRPORT/AIRCRAFT", "AIRCRAFT", 
           "AIRPORT PARKING LOT", "AIRPORT EXTERIOR - NON-SECURE AREA", "AIRPORT TERMINAL LOWER LEVEL - SECURE AREA", 
           "AIRPORT BUILDING NON-TERMINAL - SECURE AREA", "AIRPORT EXTERIOR - SECURE AREA", "AIRPORT TRANSPORTATION SYSTEM (ATS)",
           "AIRPORT TERMINAL MEZZANINE - NON-SECURE AREA", "AIRPORT TERMINAL UPPER LEVEL - NON-SECURE AREA"]
    df["location_description"].replace(air, "AIRPORT", inplace = True)

    #Combine all values related to School/Day Care
    school = ["SCHOOL, PUBLIC, BUILDING", "SCHOOL, PUBLIC, GROUNDS", "SCHOOL - PUBLIC BUILDING", "SCHOOL - PUBLIC GROUNDS", 
              "PUBLIC HIGH SCHOOL", "SCHOOL, PRIVATE, BUILDING", "SCHOOL, PRIVATE, GROUNDS", "SCHOOL - PRIVATE BUILDING", 
              "SCHOOL - PRIVATE GROUNDS", "SCHOOL YARD", "DAY CARE CENTER"]
    df["location_description"].replace(school, "SCHOOL", inplace = True)

    #Combine all values related to Bar/Tavern/Restaurant
    restaurant = ["BAR OR TAVERN", "TAVERN"]
    df["location_description"].replace(restaurant, "RESTAURANT", inplace = True)

    #Combine all values related to stores
    store = ["SMALL RETAIL STORE", "DEPARTMENT STORE", "GROCERY FOOD STORE", "CONVENIENCE STORE", "DRUG STORE", 
             "TAVERN/LIQUOR STORE", "CLEANING STORE", "APPLIANCE STORE", "TAVERN / LIQUOR STORE", "RETAIL STORE", 
             "LIQUOR STORE", "NEWSSTAND", "PAWN SHOP", "MOVIE HOUSE/THEATER", "MOVIE HOUSE / THEATER", "BARBERSHOP", 
             "CAR WASH", "COIN OPERATED MACHINE", "BOWLING ALLEY","KENNEL", "BARBER SHOP/BEAUTY SALON", "CLEANERS/LAUNDROMAT"]
    df["location_description"].replace(store, "STORE", inplace = True)

    #Combine all values related to hospitals or nursing homes
    hos = ["HOSPITAL BUILDING/GROUNDS", "HOSPITAL BUILDING / GROUNDS", "ANIMAL HOSPITAL", "NURSING HOME/RETIREMENT HOME", 
           "NURSING / RETIREMENT HOME", "NURSING HOME"]
    df["location_description"].replace(hos, "HOSPITAL", inplace = True)

    #Combine all values related to hotels/motels
    hotel = ["HOTEL/MOTEL", "HOTEL / MOTEL", "MOTEL"]
    df["location_description"].replace(hotel, "HOTEL", inplace = True)

    #Combine all values related to offices
    office = ["COMMERCIAL / BUSINESS OFFICE", "MEDICAL/DENTAL OFFICE", "MEDICAL / DENTAL OFFICE"]
    df["location_description"].replace(office, "OFFICE", inplace = True)

    #Combine all values related to buildings
    building = ["ABANDONED BUILDING", "GOVERNMENT BUILDING/PROPERTY", "FACTORY/MANUFACTURING BUILDING", 
                "GOVERNMENT BUILDING / PROPERTY", "FEDERAL BUILDING", "FACTORY / MANUFACTURING BUILDING", "GOVERNMENT BUILDING",
               "WAREHOUSE", "ELEVATOR", "YMCA"]
    df["location_description"].replace(building, "BUILDING", inplace = True)

    #Combine all values related to Chicago Housing Authority (CHA)
    cha = ["CHA APARTMENT", "CHA HALLWAY/STAIRWELL/ELEVATOR", "CHA HALLWAY / STAIRWELL / ELEVATOR", "CHA GROUNDS", 
           "CHA PLAY LOT", "CHA HALLWAY", "CHA ELEVATOR"]
    df["location_description"].replace(cha, "CHA", inplace = True)

    #Combine all values related to Church
    church = ["CHURCH/SYNAGOGUE/PLACE OF WORSHIP", "CHURCH / SYNAGOGUE / PLACE OF WORSHIP", "CHURCH PROPERTY"]
    df["location_description"].replace(church, "CHURCH", inplace = True)

    #Combine all other miscellaneous vaules
    other = ["OTHER (SPECIFY)", "OTHER RAILROAD PROP / TRAIN DEPOT", "OTHER COMMERCIAL TRANSPORTATION", 
             "OTHER RAILROAD PROPERTY / TRAIN DEPOT", "CEMETARY", "FARM", "HORSE STABLE", "RAILROAD PROPERTY",
             "WOODED AREA", "SEWER"]
    df["location_description"].replace(other, "OTHER", inplace = True)
    


In [56]:
#Clean location description variable
location_cleaner(chicago_train)
chicago_train["location_description"].value_counts()

STREET               639494
RESIDENCE            634663
APARTMENT            366599
SIDEWALK             277648
STORE                174996
OTHER                110876
PARKING               88353
RESTAURANT            74485
SCHOOL                64492
ALLEY                 59376
VEHICLE               57290
CTA                   57042
GAS STATION           33424
PARK PROPERTY         21477
BANK                  20019
OFFICE                19074
HOSPITAL              18387
BUILDING              16748
HOTEL                 13006
AIRPORT               12855
VACANT                11034
POLICE                10114
CHA                    9771
SPORTS                 7598
CHURCH                 6033
CONSTRUCTION SITE      3427
LIBRARY                2511
COLLEGE                2469
WATER                  1593
Name: location_description, dtype: int64

We reduced the number of categories from 197 to only 29. 

### Cleaning Date Variable

In [66]:
#Check for missing values
chicago_train["date"].isnull().sum()

0

We'll split the date variable into months and hour to look at the difference in crimes by time

In [297]:
def crime_month(df):
    """
    This function creates a new column that contains the month the crime occurred.
    
    df: DataFrame that has a column for the date the crime occurred.
    
    returns: None
    
    """
    df["Month"] = df.loc[:, "date"].dt.month


In [300]:
%%capture --no-stdout
#Add the Month variable
crime_month(chicago_train)

In [86]:
#Look at the total number of crimes ordered by month
chicago_train["Month"].value_counts()\
                      .reset_index()\
                      .rename(columns={"Month": "Count", "index":"Month"})\
                      .sort_values(by="Month")

Month   Count
8       1  222888
11      2  193914
6       3  227062
7       4  224360
3       5  251136
2       6  252769
0       7  264067
1       8  261356
4       9  241026
5      10  240374
9      11  219026
10     12  216876

In [82]:
def crime_hour(df):
    """
    This function creates a new column that contains the hour the crime occurred.
    
    df: DataFrame that has a column for the date the crime occurred.
    
    returns: None
    
    """
    df["Hour"] = df.loc[:, "date"].dt.hour

In [84]:
%%capture --no-stdout
#Add the Hour variable
crime_hour(chicago_train)

In [87]:
#Look at the total number of crimes ordered by hour
chicago_train["Hour"].value_counts() \
                     .reset_index() \
                     .rename(columns={"Hour": "Count", "index":"Hour"}) \
                     .sort_values(by="Hour")

Hour   Count
9      0  142456
17     1   85249
18     2   72964
20     3   60168
22     4   46156
23     5   40027
21     6   47901
19     7   66662
16     8   98523
12     9  130410
14    10  125435
13    11  129741
0     12  166232
11    13  134993
8     14  143440
3     15  154306
6     16  147882
5     17  148298
2     18  157933
1     19  160215
4     20  153122
7     21  143659
10    22  142061
15    23  117021

### Clean Domestic Variable

In [88]:
#Check for missing values
chicago_train["domestic"].isnull().sum()

0

In [89]:
#The data type says Boolean. Check that there are only two values i.e True and False
chicago_train["domestic"].value_counts()

False    2376026
True      438828
Name: domestic, dtype: Int64

Thus, there's nothing we need to clean for the domestic variable

### Clean Community Name and District Variables

Since the district variable was created by merging with community areas, cleaning community name will clean district

In [90]:
#Check the number of missing values for community names
chicago_train["community_name"].isnull().sum()

218

In [96]:
#Check the number of missing values for district
chicago_train["district_name"].isnull().sum()

218

In [92]:
#Take a look at the data filtered for missing community name values
mask = chicago_train["community_name"].isnull()
chicago_train.loc[mask].head()

unique_key                      date     primary_type  \
225896     8487665 2012-02-19 18:54:00+00:00  CRIMINAL DAMAGE   
225897     8150814 2011-07-07 14:00:00+00:00          ASSAULT   
225898     8197932 2011-08-04 17:00:00+00:00    OTHER OFFENSE   
225899     8274387 2011-09-21 12:45:00+00:00          BATTERY   
225900     8039300 2011-04-28 22:50:00+00:00          BATTERY   

                 location_description  arrest  domestic  year   latitude  \
225896               RESIDENCE-GARAGE   False     False  2012  41.799090   
225897  RESIDENTIAL YARD (FRONT/BACK)   False      True  2011  41.798980   
225898                      RESIDENCE   False     False  2011  41.798980   
225899                   CTA PLATFORM   False     False  2011  41.970515   
225900                      RESIDENCE   False      True  2011  41.798980   

        longitude district_name community_name  Month  Hour  
225896 -87.790549           NaN            NaN      2    18  
225897 -87.790547           NaN            NaN      7    14  
225898 -87.790547           NaN            NaN      8    17  
225899 -87.836651           NaN            NaN      9    12  
225900 -87.790547           NaN            NaN      4    22

The date variable shows that the first few missing values are from the beginning of the observations. Let's examine this further.

In [93]:
chicago_train[mask]["year"].value_counts()

2011    186
2012     27
2013      4
2020      1
Name: year, dtype: Int64

Since nearly all missing community names are from the beginning of the data set, we can subjectively decide to drop them. We'll do this in a function.

In [101]:
def community_cleaner(df):
    """
    This function cleans the community_name column by dropping those rows with missing values.
    
    df: DataFrame that contains the community_name column
    
    returns: None
    """
    df.dropna(subset=["community_name"], inplace = True)


In [98]:
%%capture --no-stdout
#Clean the community names
community_cleaner(chicago_train)

In [99]:
#Check that there are no missing values for community names
chicago_train["community_name"].isnull().sum()

0

In [100]:
#Check that there are no missing values for district
chicago_train["district_name"].isnull().sum()

0

In [107]:
#Google says there are 77 community areas in Chicago
assert 77 == len(chicago_train["community_name"].value_counts())

In [105]:
#Examine the number of crimes by community name
chicago_train["community_name"].value_counts()

AUSTIN                    175473
NEAR NORTH SIDE           106612
SOUTH SHORE                94836
NORTH LAWNDALE             90965
HUMBOLDT PARK              88954
NEAR WEST SIDE             87347
LOOP                       84337
AUBURN GRESHAM             80626
WEST TOWN                  78979
WEST ENGLEWOOD             78415
ROSELAND                   74522
GREATER GRAND CROSSING     71155
ENGLEWOOD                  70713
CHICAGO LAWN               66334
CHATHAM                    65530
LAKE VIEW                  61453
WEST GARFIELD PARK         60011
LOGAN SQUARE               54052
EAST GARFIELD PARK         53331
NEW CITY                   52561
SOUTH CHICAGO              50873
BELMONT CRAGIN             50749
SOUTH LAWNDALE             48034
WEST PULLMAN               44921
LINCOLN PARK               43550
WOODLAWN                   41312
ROGERS PARK                40911
UPTOWN                     38650
GRAND BOULEVARD            35938
WEST RIDGE                 35823
PORTAGE PA

In [103]:
#Examine the number of crimes by district
chicago_train["district_name"].value_counts()

West             710206
Southwest        402691
South            376621
Far Southeast    335263
Far North        246694
Central          209322
North            199119
Far Southwest    178381
Northwest        156339
Name: district_name, dtype: int64

### Clean Arrest Variable

In [108]:
#Check for missing values
chicago_train["arrest"].isnull().sum()

0

In [109]:
#The data type for arrest says Boolean so we only expect True or False
chicago_train["arrest"].value_counts()

False    2142310
True      672326
Name: arrest, dtype: Int64

We see that the arrest variable does not need any cleaning

Now that we have created functions to clean up each individual variable, we can create a wrapper function combining all individual functions

In [122]:
def chicago_data_cleaner(df, clean_crime = True, impute_location = True, clean_location = True, add_month = True,
                                add_hour = True, clean_community = True, verbose = False):
    """
    This function is a wrapper over various functions to clean the columns in the Chicago data set in place
    
    The parameters are boolean values to decide if each individual function should be call on the data set
    
    returns: None
    """
    if verbose: print("Cleaning Started...\n")
    if clean_crime:
        crime_type_cleaner(df)
        if verbose: print("Successfully Cleaned Primary Type")
        
    if impute_location:
        location_imputer(df)
        if verbose: print("Successfully Imputed Location")
    
    if clean_location:
        location_cleaner(df)
        if verbose: print("Successfully Cleaned Location")  
    
    if add_month:
        crime_month(df)
        if verbose: print("Successfully Added Month Column")
        
    if add_hour:
        crime_hour(df)
        if verbose: print("Successfully Added Hour Column")
        
    if clean_community:
        community_cleaner(df)
        if verbose: print("Successfully Cleaned Community\n")
        
    df.reset_index(drop = True, inplace = True)
    
    if verbose: print("Data Set Successfully Cleaned!")


In [120]:
#Original uncleaned training data
chicago_train = chicago.loc[chicago["year"] != 2021]
chicago_train.isnull().sum()

unique_key                 0
date                       0
primary_type               0
location_description    7844
arrest                     0
domestic                   0
community_area           212
year                       0
district_name            218
community_name           218
dtype: int64

In [123]:
%%capture --no-stdout
chicago_data_cleaner(chicago_train, verbose = True)

Cleaning Started...

Successfully Cleaned Primary Type
Successfully Imputed Location
Successfully Cleaned Location
Successfully Added Month Column
Successfully Added Hour Column
Successfully Cleaned Community

Data Set Successfully Cleaned!


In [124]:
chicago_train.isnull().sum()

unique_key              0
date                    0
primary_type            0
location_description    0
arrest                  0
domestic                0
community_area          0
year                    0
district_name           0
community_name          0
Month                   0
Hour                    0
dtype: int64